drivers: adc: SAM0: Enable automatic result devision when oversampling

guenzel-kinexon opened the pull request:
The SAM0 ADC has an internal register to divide the accumulated results of oversampling the ADC by the amount of samples collected. This value has to be set by the driver and was missing.

Fixes  #74607

decsny commented:
please address the compliance failures

guenzel-kinexon commented:
I'll try to get it done as soon as possible

guenzel-kinexon commented:
> please address the compliance failures

done

nandojve commented on drivers/adc/adc_sam0.c:
```diff
@@ -17,7 +17,7 @@ LOG_MODULE_REGISTER(adc_sam0, CONFIG_ADC_LOG_LEVEL);
 #include "adc_context.h"
 
-#if defined(CONFIG_SOC_SERIES_SAMD21) || defined(CONFIG_SOC_SERIES_SAMR21) || \
+#if defined(CONFIG_SOC_SERIES_SAMD21) || defined(CONFIG_SOC_SERIES_SAMR21) ||                      \
```
no spaces/tab before **\\** at EOF

guenzel-kinexon replied:
reverted

nandojve commented on drivers/adc/adc_sam0.c:
```diff
@@ -67,8 +67,7 @@ static void wait_synchronization(Adc *const adc)
 
-static int adc_sam0_acquisition_to_clocks(const struct device *dev,
-					  uint16_t acquisition_time)
+static int adc_sam0_acquisition_to_clocks(const struct device *dev, uint16_t acquisition_time)
```
do not reformat here

guenzel-kinexon replied:
reverted

nandojve commented on drivers/adc/adc_sam0.c:
```diff
@@ -81,12 +80,10 @@ static int adc_sam0_acquisition_to_clocks(const struct device *dev,
 	case ADC_ACQ_TIME_MICROSECONDS:
-		scaled_acq = (uint64_t)ADC_ACQ_TIME_VALUE(acquisition_time) *
-			     1000000U;
+		scaled_acq = (uint64_t)ADC_ACQ_TIME_VALUE(acquisition_time) * 1000000U;
```
The correct style is like below if you want to fix
```c
scaled_acq = (uint64_t)ADC_ACQ_TIME_VALUE(acquisition_time)
           * 1000000U;
```

guenzel-kinexon replied:
reverted

nandojve commented on drivers/adc/adc_sam0.c:
```diff
@@ -121,8 +118,7 @@ static int adc_sam0_channel_setup(const struct device *dev,
 	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
-		retval = adc_sam0_acquisition_to_clocks(dev,
-							channel_cfg->acquisition_time);
+		retval = adc_sam0_acquisition_to_clocks(dev, channel_cfg->acquisition_time);
```
roolback this

guenzel-kinexon replied:
reverted

nandojve commented on drivers/adc/adc_sam0.c:
```diff
@@ -279,25 +273,21 @@ static void adc_sam0_start_conversion(const struct device *dev)
 
-static void adc_context_update_buffer_pointer(struct adc_context *ctx,
-					      bool repeat_sampling)
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat_sampling)
```
roolback this

guenzel-kinexon replied:
reverted

nandojve commented on drivers/adc/adc_sam0.c:
```diff
@@ -328,6 +317,18 @@ static int start_read(const struct device *dev,
 	}
 
 	adc->AVGCTRL.reg = ADC_AVGCTRL_SAMPLENUM(sequence->oversampling);
+	switch (sequence->oversampling) {
```
You shoud create an enum to store the `oversampling` value and do not add a default on the switch. Compiler will ensure that all valid entries defined on the enum will be on the switch case.

guenzel-kinexon replied:
The oversampling value is a numeric value from the DT which could ofc. be converted to an enum but I feel it would be an overkill to cast it just for the switch statement. 

The default is modeled after how the peripheral works. All oversampling values above 4 have a constant adjustment value. An argument could be made that maybe  a 
```c
if (oversampling < 4) {} else {} 
```
would be more suited here.

nandojve replied:
Hi @guenzel-kinexon ,

Let's forget for a moment the implementation details and focus on the fix.

I was looking on the datasheet and it is not clear to me what you want to achieve. Could you explain in details what you want to achieve?

I see main sections (below) to explain how to use those bits. They should be combined with AVGCTRL.SAMPLENUM[3:0] but you are not changing then.
45.6.2.9 Accumulation
45.6.2.10 Averaging
45.6.2.11 Oversampling and Decimation

guenzel-kinexon replied:
Ultimately I want to use the peripheral to do hardware-averaging. To make this work the registers for accumulation as well as averaging have to be set.

The accumulation is already done in the driver (line 319, right above the switch statement). What is missing, is the averaging configuration, which I added in this PR.
As described in Accumulation section of the datasheet (chapter 33.6.6 for the SAMD21 datasheet), up to 16 samples the hardware does no automatic division. For 32 samples it already adds a factor but that is expected that the user already set the correct factor in the ADJRES field (as described in the Averaging chapter).

The datasheet is not super clear about this imho. but with trial&error on real hardware I figured it out to be a manual+automatic scheme requiring the driver to set the adjustment factor from 0..4 for averaging up to 16 samples and then let the hardware handle additional division for averaging factors beyond that.

nandojve replied:
Ok, I see now. Thank you for the fix.

Yes, you can use` if (oversampling < 4) {} else {} `

guenzel-kinexon replied:
done

nandojve reviewed (changes requested):
Hi @guenzel-kinexon,

You are doing far more than applying the fix. On Atmel we use 80 chars.
Split you PR doing first the coding formatting that you want.

Do no use clang!

Then add your change.

guenzel-kinexon commented:
> Hi @guenzel-kinexon,
> 
> You are doing far more than applying the fix. On Atmel we use 80 chars. Split you PR doing first the coding formatting that you want.
> 
> Do no use clang!
> 
> Then add your change.

Doesn't the contribution guidelines enforce a Clang formatting of the code? I am a bit lost which formatting rules to follow here

pdgendt commented:
> Doesn't the contribution guidelines enforce a Clang formatting of the code? I am a bit lost which formatting rules to follow here

No clang format is informative, not blocking.

guenzel-kinexon commented:
I will try to fix it up asap.

guenzel-kinexon committed:
driver: adjust res when oversampling

The SAM0 ADC has an internal register to divide the accumulated results
of oversampling the ADC by the amount of samples collected. This value
has to be set by the driver and was missing.

Signed-off-by: Patrick Günzel <patrick.guenzel@kinexon.com>
